### DISCUSSION TITLE
[Question][python ]undefined reference to `_Py_Dealloc'

### DISCUSSION THREAD

**Discussion by jqhr** [pkgxdev/pkgx#874] [Help]
## Environment
pkgx 1.1.0
os: debian 12
### Error description
In my poetry project, the python must be lower than 3.11, so I use pkgx install it
```bash
env +python.org@3.10
```
And when I run those command
```bash
poetry install
```
It was going wrong.
![图片](https://github.com/pkgxdev/pkgx/assets/23305067/64e304ac-3ed4-4b8a-b758-e1d5e8c39bf7)

The wrong package was "hnswlib", which must be compiled with g++ and python.
The g++ command is 
```bash
g++ -pthread -shared -pie -pie build/temp.linux-x86_64-cpython-310/./python_bindings/bindings.o -L/home/jqhr/.pkgx/python.org/v3.10.13/lib -o build/lib.linux-x86_64-cpython-310/hnswlib.cpython-310-x86_64-linux-gnu.so -fopenmp -pthread
```
And *ld* error.
![图片](https://github.com/pkgxdev/pkgx/assets/23305067/966a507f-a253-48f9-916c-0b22d98d39fd)
The *LD_LIBRARY_PATH* is
![图片](https://github.com/pkgxdev/pkgx/assets/23305067/21d17b2b-5c27-4007-9318-d512d136b54e)
I saw it include with pkgx's python path.
What can I do ?

**Comment by jqhr** [comment]
I think this question is about *development* version of python.
For ubuntu and debian, they both have "python-dev" package which include some development .so files.
But for pkgx,
How to install python with dev package?

**Reply by ysgdw** [reply]
Dev Packages generally refer to libraries, modules, or toolsets that are used to develop software. Python usually uses pip (Python package Manager) to install these packages. However, sometimes these packages require some dependencies to be installed in the system in advance. First of all, you want to install Python dev package, you need to ensure that installed on your system first Python development environment. You can then install Python's dev package with the following command: Debian/Ubuntu Linux sudo apt to get the update sudo apt - get the install python3 - dev # for Python 3 x Red Hat Linux/CentOS sudo yum Install python3 devel - # for Python 3. Use Homebrew macOS x installation: The brew install python3 # for Python 3. X Windows in Windows, can be downloaded by Python developers package (Python Developer 's Guide) to obtain the required development tools and libraries. In addition, if you use a virtual environment (virtualenv) to manage the Python environment, you can install in the virtual environment dev package: # in the virtual environment to install Python dev package PIP install Python - dev - the tools, please note that the Python - dev - tools may not be a real package name, other name may actually be, depending on the tools and libraries you need. In general, on most systems, you just need to install python3-dev or python3-devel or use Homebrew to install Python.

**Reply by jqhr** [reply]
I know that.
But I need to install multiple version of python3. so I cannot install python3-dev or python3-devel
Pkgx support multiple python version.
I just want to know how to use dev package

**Reply by jhheider** [reply]
`pkgx` doesn't seperate headers from regular packages. everything you need should be in the `python.org` package for a give version. so, if you have a dev environment with `python.org` added to it, your CPATH should include its `/include` dir, your LIBRARY_PATH should include it's /lib, etc.

**Comment by jhheider** [comment]
based on the `_Py_Dealloc` error and some quick googling, it looks like maybe you have some intermediate objects that might have been built with different python versions (https://github.com/namhyung/uftrace/issues/1190). does `make clean` help? otherwise, I might try a different `python` version and see if that's the issue.

searching my `~/.pkgx` directory shows some `pyobject*` files in the `boost.org` package. is it possible your package needs the boost library?

**Reply by jqhr** [reply]
Ok, I will have a try.

**Reply by jqhr** [reply]
I have a try. I have deleted all of python package of my debian system. Only   pkgx's 3.10 left.But it still going wrong.
![图片](https://github.com/pkgxdev/pkgx/assets/23305067/73c2def6-8e8a-4cf1-b2a9-4421c9996ad8)
![图片](https://github.com/pkgxdev/pkgx/assets/23305067/dcb78b7f-aa6d-42af-a140-611c4129b365)
Maybe some other thing or config break my  compiler.

**Reply by jhheider** [reply]
have you tried adding `boost.org: '*'` to your build dependencies? https://github.com/pybind/pybind11 seems to be a bridge intended to work with boost, if i'm reading it correctly.

**Reply by jqhr** [reply]
I have use `env +boost.org`, but the problem remains.
I have tried to remove `python`, `python3-dev`, `direnv`,  and not work.
This is very strange.

**Reply by jhheider** [reply]
Have you added it to the package.yml? I don't recall, offhand, if the dev environment passes through to package builds, but it shouldn't.

**Comment by khali245** [comment]
try to update your version it might not compatible with the package.
